src/com/google/javascript/jscomp/CompilationLevel.java

```diff
@@ -187,9 +187,11 @@ public void setTypeBasedOptimizationOptions(CompilerOptions options) {
         options.disambiguateProperties = true;
         options.ambiguateProperties = true;
         options.inlineProperties = true;
+        options.useTypesForOptimization = true;
         break;
       case SIMPLE_OPTIMIZATIONS:
         // TODO(johnlenz): enable peephole type based optimization.
+        options.useTypesForOptimization = true;
```

Comment:
The philosophy of simple-mode is to work soundly on most code, with very minimal restrictions (using around reflection). We shouldn't be using type-based optimizations in simple mode.